feat: use Fluent Bit for trial logs [DET-4178]

dzhu opened the pull request:
## Description

This PR moves trial logs to being sent through a Fluent Bit container running alongside each agent in the cluster. For the moment, doing so mostly just adds complexity, but the idea is to later add support for Elastic as a separate log backend, which will be made easier by the use of Fluent Bit.

The agent internally manages the lifecycle of the Fluent Bit container in order to keep administration simple and ensure that the daemon is started with the necessary configuration. The agent automatically configures Fluent Bit to connect back to the master and fill in the new fields being used for filtering. To simplify the required networking, the logs that the agent itself inserts are sent to the master rather than through Fluent Bit.

The master gains a new endpoint to accept POSTed trial logs, which it sends over to the trial logger actor. Only trial logs are sent through this new path; in the long term, that should also be unified with commands and other tasks, but those aren't the same currently anyway, so that's being left out of this project.

Features/todo:
- [x] start/stop Fluent Bit on agent startup/shutdown
- [x] handle existing Fluent Bit on startup
- [x] populate the new fields in the log records
- [x] send trial logs through Fluent and others through current mechanisms
- [x] port over agent-inserted meta logs
- [x] add new filtering fields for master-inserted logs
- [x] allow the agent to connect to Fluent while itself running inside Docker
- [x] improve compatibility by adding a new field for the log body
- [x] document installation requirements (new image to pre-pull for airgapped clusters)

## Test Plan

- [x] run trials and check that logs show up with reasonable formatting
- [x] run commands and check that logs show up as before
- [x] check logs of the Fluent Bit container while things are happening
- [x] start the agent with a Fluent Bit container already running
- [x] run a trial without this PR, migrate to the new version, check that logs still show up reasonably

## Commentary

Also includes the schema change PR (#1373) and some debugging code.

stoksc commented on agent/internal/fluent.go:
```diff
@@ -0,0 +1,263 @@
+	return port, createResponse.ID, nil
+}
+
+func (a *agent) killLoggingContainer() error {
+	c, err := client.NewClientWithOpts(client.FromEnv, client.WithVersion("1.40"))
+	if err != nil {
+		return errors.Wrap(err, "error connecting to Docker daemon")
+	}
+
+	defer func() {
+		if cErr := c.Close(); cErr != nil {
+			logrus.WithError(cErr).Warn("failed to close Docker connection")
+		}
+	}()
+
+	return c.ContainerKill(context.Background(), a.fluentContainerID, "KILL")
+}
```
same comment here (and elsewhere) for pulling this docker machinery into a package. thoughts?

stoksc commented on agent/internal/container.go:
```diff
@@ -20,25 +23,36 @@ type containerActor struct {
 		pull := pullImage{PullSpec: c.spec.PullSpec, Name: c.spec.RunSpec.ContainerConfig.Image}
 		ctx.Tell(c.docker, pull)
 
+		for _, env := range c.spec.RunSpec.ContainerConfig.Env {
```
didn't you scheme up a better way to do this?

stoksc reviewed (commented):
have a bunch of questions but over all looks really good. should be a fast review.

dzhu commented on agent/internal/fluent.go:
```diff
@@ -0,0 +1,263 @@
+		}
+	}
+
+	// Pull the image.
+	_, _, err = c.ImageInspectWithRaw(context.Background(), imageName)
+	switch {
+	case err == nil:
+		// No error means the image is present; do nothing.
+	case client.IsErrNotFound(err):
+		// This error means the call to Docker went fine but the image doesn't exist; pull it.
+		pullResponse, pErr := c.ImagePull(context.Background(), imageName, types.ImagePullOptions{})
+		if pErr != nil {
+			return 0, "", pErr
+		}
+		if _, pErr = ioutil.ReadAll(pullResponse); pErr != nil {
+			return 0, "", pErr
+		}
+		if pErr = pullResponse.Close(); pErr != nil {
+			return 0, "", pErr
+		}
+	default:
+		// Something unexpected happened; propagate the error.
+		return 0, "", errors.Wrap(err, "failed to pull logging image")
+	}
```
Sounds reasonable, though I'd be inclined to do that separately (assuming that you would also want to move some of the existing Docker stuff over as well).

dzhu commented:
Okay, I made the Fluent handling into its own actor, updated the trial ID/environment variable handling, and switched the Fluent config from CLI arguments to a constructed config file.

stoksc commented on agent/internal/fluent.go:
```diff
@@ -0,0 +1,425 @@
+
+// getDockerNetwork returns the name of one of the Docker networks attached to the container that
+// this process is in. It returns an empty string if it does not seem to be running inside Docker.
+func getDockerNetwork() (string, error) {
```
thought: reading all this and the other fluent code, and thinking ahead to agent logs, I'm really beginning to question having the agent manage the fluent container again. how are you feeling about this?

stoksc replied:
The main reason I'm starting to get worried about this is thinking about the UX if the agent starts fluentd, the agent logs to fluentd, _and_ fluentd doesn't start for some reason. I think that if docker hub is down or they're in an airgapped env and forgot to prepull them image, you're suddenly in a very hard situation to debug. Whereas if you start it on deploy or agent creation, in theory the deploy script could fail violently and obviously

stoksc commented on docs/how-to/installation/docker.txt:
```diff
@@ -170,6 +170,17 @@ Note that if an agent container is running on the same machine as the
+Determined internally makes use of `Fluent Bit
+<https://fluentbit.io>`__; the agent requires the
+``fluent/fluent-bit:1.5`` Docker image to be present at runtime.
+Normally, the agent will pull the Fluent Bit image automatically, but,
+if the machines in the cluster are not able to access Docker Hub, the
```
fun grammar style question (i'm bad at this). I would've done
```suggestion
Normally, the agent will pull the Fluent Bit image automatically, but
if the machines in the cluster are not able to access Docker Hub the
```
since the if clause affects the meaning of what comes after it and isn't just additive.
Why did you pick this?

dzhu replied:
The second comma is a pretty clear win for me. Not entirely mandatory, but correct (consider the standalone "if A, then B") and, as a matter of style, I think too much of the sentence runs together without it. I think you're right that the first one doesn't belong, though. Even I can't be perfect every time. :wink:

stoksc reviewed (commented):
looks great so far (though I am starting to get a little bit of a bad feeling about the agent managing fluent), just had another round of mostly questions. going to give it a test on macos now.

stoksc commented on agent/internal/fluent.go:
```diff
@@ -0,0 +1,425 @@
+// an appropriate configuration.
+func fluentConfig(opts Options) ([]string, archive.Archive, error) {
+	const luaPath = "/run/determined/fluent/tonumber.lua"
+	const configPath = "/ru/determined/fluent/fluent.conf"
```
```suggestion
	const configPath = "/run/determined/fluent/fluent.conf"
```

stoksc commented on agent/internal/fluent.go:
```diff
@@ -0,0 +1,439 @@
+		return err
+	}
+	for _, cont := range containers {
+		ctx.Log().Infof("killing found Fluent Bit container %s", cont.ID)
+		if err = c.ContainerKill(context.Background(), cont.ID, "KILL"); err != nil {
+			return errors.Wrap(err, "failed to kill existing Fluent Bit container")
```
non-blocking: the messages still are fluent specific.

stoksc commented on agent/internal/fluent.go:
```diff
@@ -0,0 +1,439 @@
+  Format regex
+  # Parse out something that looks like a log level at the start of the line (e.g., "INFO: xxx"); if
+  # nothing is found, return an empty log level and the full message back.
+  Regex ((?<level>^(DEBUG|INFO|WARNING|ERROR|CRITICAL)): |(?<level>))(?<log>.*)
```
non-blocking: are we going to include rank id?

stoksc commented on agent/internal/fluent.go:
```diff
@@ -0,0 +1,439 @@
+	}
+
+	container, _ := c.ContainerInspect(context.Background(), createResponse.ID)
+	portStr := container.NetworkSettings.Ports[exposedPort][0].HostPort
```
Still hitting an issue here on macos - we can look into it monday.

stoksc reviewed (commented):
still hitting an issue with the agent in hostmode but otherwise lgtm.

stoksc commented on agent/internal/fluent.go:
```diff
@@ -0,0 +1,439 @@
+			Cmd:          fluentArgs,
+			ExposedPorts: nat.PortSet{exposedPort: struct{}{}},
+		},
+		&container.HostConfig{
```
blocking: I think we should still be starting with --restart-policy on-failure

dzhu replied:
Leaving this off for now, per offline discussion (still sounds desirable but would have other annoying effects that we don't want to block on right now).

stoksc commented on agent/internal/fluent.go:
```diff
@@ -0,0 +1,439 @@
+			ExposedPorts: nat.PortSet{exposedPort: struct{}{}},
+		},
+		&container.HostConfig{
+			AutoRemove:  true,
```
non-blocking: I think we should also start the container with some reasonable limits for `HostConfig.Resources`

stoksc commented on agent/internal/agent.go:
```diff
@@ -123,6 +127,10 @@ func (a *agent) Receive(ctx *actor.Context) error {
 		a.handleAPIRequest(ctx, msg)
 
 	case actor.PostStop:
+		if err := a.fluent.StopAndAwaitTermination(); err != nil {
```
blocking: one last minor nit, if we're stopping because we failed to start fluent, this deref's a nil pointer and leads to a confusing-ish stack trace.

stoksc commented on agent/internal/fluent.go:
```diff
@@ -0,0 +1,450 @@
+// this process is in. It returns an empty string if it does not seem to be running inside Docker or
+// is in Docker but on the host network, which should be treated equivalently.
+func getDockerNetwork(docker *client.Client) (string, error) {
+	f, err := os.Open("/proc/self/cgroup")
+	if err != nil {
+		return "", errors.Wrap(err, "error opening cgroup file")
+	}
```
blocking: I think we need a switch case here to check if it's an `ErrNotFound`, otherwise outside of docker and not on linux (on mac) we'll fail here. I think general it's safe to assume that: IF we can't find this file OR this file doesn't contain the docker cgroup THEN we're not in a container.

stoksc commented on agent/internal/fluent.go:
```diff
@@ -0,0 +1,450 @@
+	}
+	for _, cont := range containers {
+		ctx.Log().WithFields(logrus.Fields{"name": name, "id": cont.ID}).Infof("killing Docker container")
+		if err = killContainer(ctx, docker, cont.ID); err != nil {
```
blocking: I think this is a race condition. Peeked at the docker daemon code and kill just sends a signal and returns but doesn't wait for anything to happen. In ~10 runs once I did actually see where we tried to create the container but it hadn't fulled stopped and been removed yet.
```
determined-agent   | ERRO[2020-10-28T10:07:15-04:00] error while actor was running                 error="failed to start Fluent daemon: Error response from daemon: Conflict. The container name \"/determined-fluent\" is already in use by container \"64ffc0be5858bef1612e40ef11f05978946d8292519179ec742ae1274272d97f\". You have to remove (or rename) that container to be able to reuse that name." id=agent system=Bradleys-MBP.attlocal.net type=agent
```
I think we need to throw a `c.ContainerWait()` behind all of these.

stoksc replied:
Sorry I'm leaving these all 1 by 1. As I'm writing the elastic stuff, I'm getting a lot more familiar with the code than I was able to get from review.

stoksc commented on agent/internal/fluent.go:
```diff
@@ -0,0 +1,476 @@
+		if opts.Security.TLS.MasterCert != "" {
+			outputConfig += "  tls.ca_file " + masterCertPath + "\n"
+
+			certBytes, cErr := ioutil.ReadFile(opts.Security.TLS.MasterCert)
```
comment: may be nice to add a "resolve" to the agent config and write a little package like the check package to do things like load the cert in and etc.. there's one forming in the master, just had to modify it to load the elastic cert so the master can send it to the agent.

dzhu replied:
Hmm, interesting, but I don't feel like it's worth it yet. Maybe after this PR.

stoksc commented on agent/internal/fluent.go:
```diff
@@ -0,0 +1,476 @@
+	return args, files, nil
+}
+
+func killContainerByName(ctx *actor.Context, docker *client.Client, name string) error {
```
non-blocking: maybe `removeContainerByName` now?

stoksc commented on docs/how-to/installation/docker.txt:
```diff
@@ -170,6 +170,17 @@ Note that if an agent container is running on the same machine as the
+Determined can run. In order to specify a different image to use for
+running Fluent Bit (generally to make use of a custom Docker
+registry---the image should not normally need to be changed otherwise),
+use the agent's ``--fluent-logging-image`` command-line option or
```
question: how are people using dynamic agents going to set this?

stoksc reviewed (approved):
lgtm. this definitely was a lot more work than originally anticipated. it's a little unfortunate how hairy all the networking stuff got, but we the combinations of options we have to support I don't exactly see a way out.

I think once the agents send their logs to the same place, we could do some fancy stuff to route agent-inserted-trial-logs there and the agent wouldn't need to connect to this fluent instance directly (and that's where most of the complexity is coming from).. I guess I alternative to all this, then, would be to just send the agent inserted logs back to the master through the endpoint directly until then - but I'm not willing to bet on its simplicity at this point though, haha.

stoksc reviewed (approved):
lgtm! it looks much better like this. wonderful.
